CMake: Modernize DEAL_II_PACKAGE_HANDLE macro and clean up Find*.cmake files

tamiko opened the pull request:
This pull request changes the DEAL_II_PACKAGE_HANDLE macro:

    * simplify logic by removing unnecessary booleans

    * use temporary variables to accumulate lists/string and override
     canonical names at end

This allows to use "canonical" names such as "FEATURE_LIBRARIES" already in
the call to DEAL_II_PACKAGE_HANDLE by clearing these variables at the end
and not at the beginning of the DEAL_II_PACKAGE_HANDLE macro.

tamiko commented:
I have manually verified that this still produces the same `detailed.log` output than the current master branch with one notable change: Empty variables are now removed from `detailed.log`.

tjhei commented on CMakeLists.txt:
```diff
@@ -33,14 +33,14 @@
 MESSAGE(STATUS "")
 
-CMAKE_MINIMUM_REQUIRED(VERSION 3.1.0)
+CMAKE_MINIMUM_REQUIRED(VERSION 3.3.0)
```
we currently require 3.1 according to https://www.dealii.org/developer/readme.html
Do you need 3.3? If yes then we need to adjust.

tamiko replied:
Yes. I chose to simply bump to 3.3.0. It will make some cleanup of the CMake code I plan to do easier - and version 3.3.0 is reasonably ancient (Ubuntu 16.04 LTS already shipped 3.5.0). I have adjusted the readme and the rest of the documentation to state `3.3.0`.

tjhei commented on cmake/macros/macro_deal_ii_package_handle.cmake:
```diff
@@ -56,102 +56,103 @@ MACRO(DEAL_II_PACKAGE_HANDLE _feature _var)
+  # Clear temporary variables
+  #
+  FOREACH(_suffix ${DEAL_II_LIST_SUFFIXES} ${DEAL_II_STRING_SUFFIXES})
+    set(_temp_${_suffix} "")
```
```suggestion
    SET(_temp_${_suffix} "")
```

tamiko replied:
Fixed!

tjhei commented on cmake/macros/macro_deal_ii_package_handle.cmake:
```diff
@@ -56,102 +56,103 @@ MACRO(DEAL_II_PACKAGE_HANDLE _feature _var)
+            AND "${_suffix}" STREQUAL "LIBRARIES")
+      LIST(APPEND _temp_${_suffix} ${_arg})
+
+    ELSEIF("${_arg}" STREQUAL "CLEAR") # FIXME
```
?

tamiko replied:
I forgot to finish refactoring. The variable `_fill_clear` is now gone and the `FIXME`s have been removed.

tjhei commented on cmake/macros/macro_deal_ii_package_handle.cmake:
```diff
@@ -56,102 +56,103 @@ MACRO(DEAL_II_PACKAGE_HANDLE _feature _var)
       MARK_AS_ADVANCED(${_arg})
-      IF(_fill_clear)
+
+      IF(_fill_clear) # FIXME
```
?

tjhei commented on cmake/macros/macro_deal_ii_package_handle.cmake:
```diff
@@ -56,102 +56,103 @@ MACRO(DEAL_II_PACKAGE_HANDLE _feature _var)
-      SET(_required TRUE)
-      SET(_fine TRUE)
+      set(_suffix "${_arg}")
+      SET(_fill_clear FALSE)
```
Maybe you can comment what these variables do?

tamiko replied:
Done.

tjhei reviewed (commented):
CMake is beautiful.

tamiko commented:
@tjhei Thanks for the thorough read! I hope I have addressed all review comments. I checked that my latest changes still produce the same `CMakeCache.txt` and `detailed.log` file on my system. But let's see what the testsuite has to say.

tjhei reviewed (approved):
Looks good. Would you mind writing a changelog entry for the cmake version change?

tamiko commented:
@tjhei Done!
